components/num-util/src/fixed_decimal.rs

```diff
@@ -373,6 +377,135 @@ impl fmt::Display for FixedDecimal {
     }
 }
 
+impl FromStr for FixedDecimal {
+    type Err = Error;
+    fn from_str(input_str: &str) -> Result<Self, Self::Err> {
+        // input_str: the input string
+        // no_sign_str: the input string when the sign is removed from it
+        let no_sign_str: &str;
+        let mut is_negative = false;
+        if &input_str[0..1] == "-" {
+            is_negative = true;
+            no_sign_str = &input_str[1..];
+        } else {
+            no_sign_str = input_str;
+        }
+        // Compute length of each string once and store it, so if you use that multiple times,
+        //  you don't compute it multiple times
+        // has_dot: shows if your input has dot in it
+        // dot_index: gives the index of dot (after removing the sign)
+        let no_sign_str_len = no_sign_str.chars().count();
+        let mut has_dot = false;
+        let mut dot_index = no_sign_str_len;
+        // The following loop computes has_dot, dot_index, and also checks to see if all
+        // characters are digits and if you have at most one dot
+        // Note: Input of format 111_123 is detected as syntax error here
+        // Note: Input starting or ending with a dot is detected as syntax error here (Ex: .123, 123.)
+        for (i, c) in no_sign_str.chars().enumerate() {
+            if c == '.' {
+                match has_dot {
+                    false => {
+                        dot_index = i;
+                        has_dot = true;
+                        if i == 0 || i == no_sign_str.len() - 1 {
+                            return Err(Error::Syntax);
+                        }
+                    }
+                    true => {
+                        return Err(Error::Syntax);
+                    }
+                }
+            } else if !c.is_digit(10) {
+                return Err(Error::Syntax);
+            }
+        }
+        // no_dot_str: shows the string when the dot is removed from it
+        let no_dot_str = no_sign_str.replace(".", "");
+        let no_dot_str_len = no_dot_str.chars().count();
```

Comment:
These are both O(N) operations.

You can get `no_dot_str_len` by doing `dot_str_len - 1`.  If it makes you feel better, you can use `debug_assert!` that the length is what you expect it to be.

You should try to avoid creating `no_dot_str`.  You can count `left_zeros` in the same loop where you figure out `dot_index`.  For `right_zeros`, you can make a reverse iterator on `no_sign_str` and just skip the character when you get to `'.'`.

Reply:
I also use `no_dot_str` later in order to make `digits_str`. I can still make dec.digits directly from `no_sign_str` (by conditioning on dot_index, left_zeros, and right_zeros), but I thought this way the code would be much easier to read. Do you think I still need to remove both `no_dot_str` and `digits_str`?

Reply:
Think of it this way.  If the input string is 50,000 bytes long, this line requires a memory allocation of 49,999 bytes.  The memory allocation is unnecessary to the algorithm, so it should be avoided.

I think you can keep your same code and just use indices into `no_sign_str` everywhere, and whenever you loop over it, check for and skip the `.`.  If it makes it easier, assume (as you're already doing) that `dot_index` equals the string length if there is no dot in the string.